gen_exports.py

```diff
@@ -0,0 +1,223 @@
+#! /usr/bin/env python3
+# -*- coding: utf-8 -`-
+"""
+Code generation script for class methods
+to be exported as public API
+"""
+import argparse
+import ast
+import astor
+import difflib
+import os
+import yapf.yapflib.yapf_api as formatter
+
+SOURCE_TREE = './trio/_core'
+YAPF_STYLE = './.style.yapf'
+
+IMPORTS = """# ***********************************************************
+# ******* WARNING: AUTOGENERATED! ALL EDITS WILL BE LOST ******
+# *************************************************************
+from ._run import GLOBAL_RUN_CONTEXT, _NO_SEND
+from ._ki import LOCALS_KEY_KI_PROTECTION_ENABLED
+
+    
+"""
+
+TEMPLATE = """locals()[LOCALS_KEY_KI_PROTECTION_ENABLED] = True
+try:
+    return GLOBAL_RUN_CONTEXT.{}.{}
+except AttributeError:
+    raise RuntimeError('must be called from async context')
+"""
+
+
+def is_function(node):
+    """Check if the AST node is either a function
+    or an async function
+    """
+    return isinstance(node, ast.FunctionDef) \
+           or isinstance(node, ast.AsyncFunctionDef)
+
+
+def is_public(node):
+    """Check if the AST node has a _public decorator
+    """
+    return is_function(node) \
+           and node.decorator_list \
+           and isinstance(node.decorator_list[-1], ast.Name) \
+           and node.decorator_list[-1].id == '_public'
+
+
+def get_public_methods(tree):
+    """ Return a list of methods marked as public.
+    The function walks the given tree and extracts
+    all objects that are functions and have a
+    doc string that starts with PUBLIC
+    """
+    methods = []
+    for node in ast.walk(tree):
+        for child in ast.iter_child_nodes(node):
+            if is_public(child):
+                child.parent = node
+                child.module_file = tree.module_file
+                child.module_path = tree.module_path
+                methods.append(child)
+    return methods
+
+
+def get_module_trees_by_dir(source_dir):
+    """ Converts a list of modules into ast module objects
+    """
+    mod_files = astor.code_to_ast.find_py_files(source_dir)
+    modules = []
+    for mod_file in mod_files:
+        if not mod_file[1].startswith('_public'):
+            module = astor.code_to_ast.parse_file(os.path.join(*mod_file))
+            module.module_path = mod_file[0]
+            module.module_file = mod_file[1]
+            modules.append(module)
+    return modules
+
+
+def get_export_modules_by_dir(source_dir):
+    """Return all modules in the given directory
+    and its sub directories which contain methods
+    marked to be public as AST module trees
+    """
+    export_modules = []
+    modules = get_module_trees_by_dir(source_dir)
+
+    for module in modules:
+        methods = get_public_methods(module)
+        if methods:
+            export_modules.append(module)
+    return export_modules
```

Comment:
This whole section feels more elaborate than we need... we know there are exactly 4 files we care about, and we could just hard-code it here. If the list ever changes, well, this lives in the source tree right next to the files, so we can just update it :-)

I guess if you prefer to do things this way it's not actually wrong? Just saying it could be simplified too.

Reply:
We currently have three major OS but that might change in the future. Someone might port it to BSD or even to REDOX. Then you just have to add your modules without worrying where they have to live. IMHO we should just leave it now that it is in place instead of reverting, unless there are major disadvantages.